src/theory/quantifiers/sygus/cegis_core_connective.h

```diff
@@ -0,0 +1,386 @@
+ * undetermined depth is in contrast to the NodeTrie (expr/node_trie.h), which
+ * assumes all data is stored at a fixed depth.
+ *
+ * Since data can be stored at any depth, we require both a d_children field
+ * and a d_data field.
+ */
+class VariadicTrie
+{
+ public:
+  /** the children of this node */
+  std::map<Node, VariadicTrie> d_children;
+  /** the data at this node */
+  Node d_data;
+  /**
+   * Add data with identifier n indexed by i, return true if data is not already
+   * stored at the node indexed by i.
+   */
+  bool add(Node n, const std::vector<Node>& i);
+  /** Is there any data in this trie that is indexed by any subset of is? */
+  bool hasSubset(const std::vector<Node>& is) const;
+};
+
+/** CegisCoreConnective
+ *
+ * A sygus module that is specialized in handling conjectures of the form:
+ * exists P. forall x. (A[x] => C(x)) ^ (C(x) => B(x))
+ * That is, conjectures with a pre/post condition but no inductive relation
+ * or other constraints. Additionally, we may have that the above conjecture
+ * has a side condition, which requires that exists x. SC[x]^C(x) is
+ * satisfiable.
+ *
+ * Two examples of this kind of sygus conjecture are abduction and
+ * interpolation.
+ *
+ * This module implements a specific algorithm for constructing solutions
+ * to this conjecture based on Boolean connectives and unsat cores, described
+ * in following. We give two variants of the algorithm, both implemented as
+ * special cases of this class. Below, let:
+ *
+ * pool(A) be a set of literals { c_1, ..., c_n } s.t. c_i => B for i=1,...,n,
+ * pts(A) : a set of points { v | A[v] is true },
+ * pool(B) : a set of literals { d_1, ..., d_m } s.t. A => d_i for i=1,...,m,
+ * pts(B) : a set of points { v | ~B[v] is true },
+ * cores(B) : a set of sets of literals { U_1, ..., U_p } s.t. for i=1,...,p:
+ * - U_i is a subset of pool(B),
+ * - A ^ U_i is unsat.
+ * We construct pool(A)/pool(B) using enumerative SyGuS, discarding the literals
+ * that do not match the criteria.
+ *
+ * Variant #1 (Interpolation)
+ *
+ * Let the synthesis conjecture be of the form
+ *   exists C. forall x. A[x] => C[x] ^ C[x] => B[x]
+ *
+ * The high level idea is we construct solutions for C of the form
+ *   c_1 OR ... OR c_n where c_i => B for each i=1,...,n, or
+ *   d_1 AND ... AND d_m where A => d_i for each i=1,...,m.
+ *
+ * while(true){
+ *   Let e_i = next_sygus_enum();
+ *   // check if e_i should be added to the pool
+ *   if e_i[v] is true for all v in pts(A)
+ *     if A => e_i
+ *       pool(B) += e_i;
+ *     else
+ *       pts(A) += { v } where { x -> v } is a model for A ^ ~e_i;
+ *   // try to construct a solution based on the pool
+ *   Let D = {}.
+ *   while
+ *     D[v] is true for some v in pts(B), and
+ *     d'[v] is false for some d' in pool(B)
+ *   {
+ *     D += { d' }
+ *     if D is false for all v in pts(B)
+ *       if D => B
+ *         return d_1 AND ... AND d_n
```

Comment:
Using `d_1... d_n` is a bit confusing here because above `d_1,...,d_m` were defined to be the elements of pool(B). In particular, `d_1` (from pool(B) ) is not always part of the returned conjunction.

How about `return x_1 AND ... AND x_k where D = {x_1 ,..., x_k}`?  or `AND(D)`?

Reply:
Ive changed this to `AND_{d in D}( d )`.

Reply:
prefect